### PR TITLE
Use polymorphic nested inputs for copy specs

### PR DESCRIPTION
**Pull request by lptr**
Previously we used the runtime TaskInputs API to register copy specs as inputs by listening for child specs being added. Once we wanted to know the actual files to copy, or the final state of inherited the properties, CopySpecResolvers were created to do that job. This code was hard to follow, somewhat redundant and inefficient in places.

This commit introduces the immutable ResolvedCopySpec and ResolvedCopySpecNode types that capture the resolved properties and sources used by both input snapshotting and the execution of the copy actions. The resolved root spec is exposed as a `@Nested` property on the AbstractCopyTask type.

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=11511080

File: subprojects/core-api/src/main/java/org/gradle/api/file/CopySpec.java [hunk 1]
```diff
@@ -218,7 +218,7 @@
     /**
      * {@inheritDoc}
      */
-    CopySpec from(Object sourcePath, Action<? super CopySpec> configureAction);
+    CopySpec from(Object sourcePath, @Nullable Action<? super CopySpec> configureAction);
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/api/file/CopySpec.java]: Yeah, I guess you're right. I was blindly reflecting how it works today, but that's probably not how it should. I wouldn't check explicitly, and fail the build, though.

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/api/file/CopySpec.java]: ...to avoid breaking the few builds that might rely on this accepting `null`.

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/PatternSet.java [hunk 2]
```diff
@@ -37,6 +38,7 @@
 /**
  * Standalone implementation of {@link PatternFilterable}.
  */
+@NonNullApi
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/PatternSet.java]: Removed.

File: subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java [hunk 3]
```diff
@@ -134,10 +70,32 @@ public boolean isSatisfiedBy(Task task) {
...
+
     protected CopySpecInternal createRootSpec() {
         Instantiator instantiator = getInstantiator();
         FileResolver fileResolver = getFileResolver();
-        return instantiator.newInstance(DefaultCopySpec.class, fileResolver, instantiator);
+        return instantiator.newInstance(RootCopySpec.class, this, fileResolver, instantiator);
+    }
+
+    @Nested
+    ResolvedCopySpecNode getResolvedRootSpec() {
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java]: I don't agree. Why would users need to add their own instances? I'm also leaning towards doing away with the whole `ResolvedCopySpec[Node]` thing, and just doing some dynamic resolution in `DefaultCopySpec` to fix the performance regressions.

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java]: ...and make `@Nested CopySpec getRootSpec()`.

File: subprojects/core/src/testFixtures/groovy/org/gradle/api/tasks/bundling/AbstractArchiveTaskTest.groovy [hunk 4]
```diff
@@ -16,13 +16,14 @@
...
 
+import org.gradle.api.tasks.AbstractCopyTask
 import org.gradle.api.tasks.AbstractCopyTaskContractTest
 
 abstract class AbstractArchiveTaskTest extends AbstractCopyTaskContractTest {
     abstract AbstractArchiveTask getArchiveTask()
 
     @Override
-    AbstractArchiveTask getTask() {
+    AbstractCopyTask getTask() {
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/testFixtures/groovy/org/gradle/api/tasks/bundling/AbstractArchiveTaskTest.groovy]: IntelliJ started to fail without it. I have no idea (hah!), nor intention to report the problem.

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=11698744

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=11718956

**Comment by blindpirate** [comment]
Are we still going to make it in `4.7RC1`? If not, please remove the milestone.

**Comment by lptr** [comment]
Thanks for the reminder, no, we are not doing this at all.
